python/lsst/afw/image/slicing.py

```diff
@@ -1,102 +1,123 @@
+            return sliceArgs, defaultOrigin
+    except TypeError:  # Arg is a scalar; return it along with the default origin.
+        return sliceArgs, defaultOrigin
 
-        imageSlice.append(s)
 
-    x, y = [_.indices(wh) for _, wh in zip(imageSlice, img.getDimensions())]
-    return lsst.geom.Box2I(lsst.geom.Point2I(x[0], y[0]), lsst.geom.Point2I(x[1] - 1, y[1] - 1)), LOCAL
+def handleNegativeIndex(index, size, origin, default):
+    """Handle negative indices passed to image accessors.
+
+    When negative indices are used in LOCAL coordinates, we interpret them as
+    relative to the upper bounds of the array, as in regular negative indexing
+    in Python.
+
+    When negative indices are used in PARENT coordinates, we interpret them as
+    actual negative pixel values.
+    """
+    if index is None:
+        assert default is not None
+        return default
+    if index < 0 and origin == LOCAL:
+        index = size + index
+    return index
+
+
+def makePointFromIndices(x, y, origin, parent):
+    """Create a Point2I from an x, y pair, correctly handling negative indices.
+    """
+    return Point2I(
+        handleNegativeIndex(x, parent.getWidth(), origin, default=None),
+        handleNegativeIndex(y, parent.getHeight(), origin, default=None)
+    )
+
+
+def makeBoxFromSlices(x, y, origin, parent):
+    """Transform a tuple of slice objects into a Box2I, correctly handling negative indices.
+    """
+    if x.step is not None or y.step is not None:
+        raise ValueError("Slices with steps are not supported in image indexing.")
+    begin = Point2I(
+        handleNegativeIndex(x.start, parent.getWidth(), origin, default=parent.getBeginX()),
+        handleNegativeIndex(y.start, parent.getHeight(), origin, default=parent.getBeginY())
+    )
+    end = Point2I(
+        handleNegativeIndex(x.stop, parent.getWidth(), origin, default=parent.getEndX()),
+        handleNegativeIndex(y.stop, parent.getHeight(), origin, default=parent.getEndY())
+    )
+    return Box2I(begin, end - begin)
+
+
+def interpretSliceArgs(sliceArgs, bboxGetter):
+    """Transform arguments to __getitem__ or __setitem__ to a standard form.
+
+    Parameters
+    ----------
+    sliceArgs : `tuple`, `Box2I`, or `Point2I`
+        Slice arguments passed directly to `__getitem__` or `__setitem__`.
+    bboxGetter : callable
+        Callable that accepts an ImageOrigin enum value and returns the
+        appropriate image bounding box.  Usually the bound getBBox method
+        of an Image, Mask, or MaskedImage object.
+
+    Returns
+    -------
+    box : `Box2I` or `None`
+        A box to use to create a subimage, or None if the slice refers to a
+        scalar.
+    index: `tuple` or `None`
+        An ``(x, y)`` tuple of integers, or None if the slice refers to a
+        box.
+    origin : `ImageOrigin`
+        Enum indicating whether to account for xy0.
```

Comment:
If you decided to overload `subset`, then returning the result as a tuple makes the caller do unnecessary processing with the result. If `slices` is either a `Box2I` or `Point2I`, then this function should just be able to return `(slices, origin)`. However if you don't overload `subset` ignore this comment.

Reply:
I'm not planning to overload `subset`; I think considering a single pixel to be a kind of subimage to be confusing, even it it isn't necessary incorrect.